### PR TITLE
support wait second grammar for select for update

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

Part of issue 
https://github.com/pingcap/tidb/issues/19579

### What is changed and how it works?

Try to support the `wait seconds` grammar in a simple way.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: ast/dml.go [hunk 1]
```diff
@@ -916,13 +926,15 @@ func (n *SelectStmt) Restore(ctx *format.RestoreCtx) error {
...
-		ctx.WritePlain(" ")
-		ctx.WriteKeyWord(n.LockTp.String())
+	if n.LockInfo != nil {
+		switch n.LockInfo.LockType {
+		case SelectLockInShareMode:
+			ctx.WriteKeyWord(" LOCK ")
+			ctx.WriteKeyWord(n.LockInfo.String())
+		case SelectLockForUpdate, SelectLockForUpdateNoWait, SelectLockForUpdateWaitN:
+			ctx.WritePlain(" ")
+			ctx.WriteKeyWord(n.LockInfo.String())
```
Comment (kennytm) [inline ast/dml.go:936]:
```suggestion
		case SelectLockForUpdate, SelectLockForUpdateNoWait:
			ctx.WritePlain(" ")
			ctx.WriteKeyWord(n.LockInfo.String())
		case SelectLockForUpdateWaitN:
			ctx.WriteKeyWord(" FOR UPDATE WAIT ")
			ctx.WritePlainf("%d", n.LockInfo.WaitSec)
```

File: ast/dml.go [hunk 2]
```diff
@@ -456,11 +458,17 @@ const (
...
 
+type SelectLockInfo struct {
+	LockType SelectLockType
+	WaitSec  uint64
+}
+
 // String implements fmt.Stringer.
-func (slt SelectLockType) String() string {
-	switch slt {
+func (n *SelectLockInfo) String() string {
```
Comment (kennytm) [inline ast/dml.go:470]: keep this `String()` a function of `SelectLockType`.

File: ast/util.go [hunk 3]
```diff
@@ -24,8 +24,12 @@ const (
...
 	switch st := node.(type) {
 	case *SelectStmt:
-		if st.LockTp == SelectLockForUpdate || st.LockTp == SelectLockForUpdateNoWait {
-			return false
+		if st.LockInfo != nil {
+			if st.LockInfo.LockType == SelectLockForUpdate ||
+				st.LockInfo.LockType == SelectLockForUpdateNoWait ||
+				st.LockInfo.LockType == SelectLockForUpdateWaitN {
+				return false
+			}
```
Comment (kennytm) [inline ast/util.go:32]: use a `switch` for so many cases.

**Comment by jackysp** [comment]
should we merge it to master? I think it is only used in PoC.

**Comment by cfzjywxk** [comment]
> should we merge it to master? I think it is only used in PoC.

I'm not quite sure which version the POC test will use, I think it's ok to merge.

**Comment by jackysp** [comment]
> > should we merge it to master? I think it is only used in PoC.
> 
> @jackysp
> I'm not quite sure which version the POC test will use, I think it's ok to merge.

OK. Please address the comments.

File: ast/util.go [hunk 4]
```diff
@@ -25,9 +25,12 @@ func IsReadOnly(node Node) bool {
...
 		if st.LockInfo != nil {
-			if st.LockInfo.LockType == SelectLockForUpdate ||
-				st.LockInfo.LockType == SelectLockForUpdateNoWait ||
-				st.LockInfo.LockType == SelectLockForUpdateWaitN {
+			switch st.LockInfo.LockType {
+			case SelectLockForUpdate:
+				return false
+			case SelectLockForUpdateNoWait:
+				return false
+			case SelectLockForUpdateWaitN:
```
Comment (kennytm) [inline ast/util.go:33]:
```suggestion
			case SelectLockForUpdate, SelectLockForUpdateNoWait, SelectLockForUpdateWaitN:
```

**Review by kennytm — COMMENTED** [review]
rest lgtm

**Review by lysu — COMMENTED** [review]
LGTM

**Review by jackysp — COMMENTED** [review]
LGTM

**Comment by jackysp** [comment]
Please resolve the conflicts.

**Review by kennytm — DISMISSED** [review]

**Comment by coocood** [comment]
LGTM

**Review by coocood — APPROVED** [review]

**Pull request merged**